components/webext-storage/src/sync/bridge.rs

```diff
@@ -46,20 +50,61 @@ impl<'a> sync15_traits::BridgedEngine for BridgedEngine<'a> {
         Err(ErrorKind::NotImplemented.into())
     }
 
-    fn store_incoming(&self, _incoming_envelopes: &[IncomingEnvelope]) -> Result<()> {
-        Err(ErrorKind::NotImplemented.into())
+    fn store_incoming(&self, incoming_envelopes: &[IncomingEnvelope]) -> Result<()> {
+        let signal = self.db.begin_interrupt_scope();
+
+        let mut incoming_payloads = Vec::with_capacity(incoming_envelopes.len());
+        for envelope in incoming_envelopes {
+            signal.err_if_interrupted()?;
+            incoming_payloads.push(envelope.payload()?);
+        }
+
+        let tx = self.db.unchecked_transaction()?;
+        stage_incoming(&tx, incoming_payloads, &signal)?;
+        tx.commit()?;
+        Ok(())
     }
 
     fn apply(&self) -> Result<ApplyResults> {
-        Err(ErrorKind::NotImplemented.into())
+        let signal = self.db.begin_interrupt_scope();
+
+        let tx = self.db.unchecked_transaction()?;
+        let incoming = get_incoming(&tx)?;
+        let actions = incoming
+            .into_iter()
+            .map(|(item, state)| (item, plan_incoming(state)))
+            .collect();
+        apply_actions(&tx, actions, &signal)?;
+        stage_outgoing(&tx)?;
+        tx.commit()?;
+
+        let outgoing = get_outgoing(&self.db, &signal)?
+            .into_iter()
+            .map(OutgoingEnvelope::from)
+            .collect::<Vec<_>>();
+        Ok(outgoing.into())
     }
 
-    fn set_uploaded(&self, _server_modified_millis: i64, _ids: &[String]) -> Result<()> {
-        Err(ErrorKind::NotImplemented.into())
+    /// TODO: This should be `SyncGuid`, not `String`.
+    fn set_uploaded(&self, _server_modified_millis: i64, ids: &[String]) -> Result<()> {
+        let signal = self.db.begin_interrupt_scope();
+        let guids = ids
+            .iter()
+            .map(|id| SyncGuid::from(id.as_str()))
+            .collect::<Vec<_>>();
+
+        let tx = self.db.unchecked_transaction()?;
+        record_uploaded(&tx, &guids, &signal)?;
+        tx.commit()?;
+
+        Ok(())
     }
 
+    /// TODO: Need a `sync_started`, so that we can create temp tables before we
```

Comment:
Ditto, except we also need a new XPIDL method that we call just before we call `storeIncoming`. This gives us nice symmetry—`syncStarted` to set everything up, `syncFinished` when we're all done.

Reply:
+1, I was just looking for that other `create_empty_sync_temp_tables()` call